tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -148,73 +148,77 @@ HloInstruction *PadOperandToMultipleOf16(absl::Span<const int64_t> batch_dims,
   return PadOperandToTargetShape(padded_shape, x);
 }
 
-// Recursively collects unary, divide, dynamic-slice, pad or multiply operands
-// of instr until an instruction with FP8 element type is reached. Returns
-// std::nullopt when no FP8 instruction is reached.
-std::optional<std::vector<HloInstruction *>> FindF8SubgraphRecursive(
+// Recursively collects unary, divide, dynamic-slice, pad, multiply or select
+// operands of instr until an instruction with FP8 element type is reached.
+// Returns an empty vector when no FP8 instruction is reached.
+std::vector<std::pair<HloInstruction *, int>> FindF8SubgraphRecursive(
     HloInstruction *instr, absl::flat_hash_set<int> &visited_instrs,
-    std::vector<HloInstruction *> subgraph) {
+    std::vector<std::pair<HloInstruction *, int>> subgraph) {
   // Avoid visiting the same instruction more than once.
   if (!visited_instrs.emplace(instr->unique_id()).second) {
-    return std::nullopt;
+    return {};
   }
-  subgraph.emplace_back(instr);
+  subgraph.emplace_back(std::make_pair(instr, 0));
   if (IsF8Type(instr)) {
     return subgraph;
-  } else {
-    if (instr->operand_count() == 1 || instr->opcode() == HloOpcode::kDivide ||
-        instr->opcode() == HloOpcode::kDynamicSlice ||
-        instr->opcode() == HloOpcode::kPad) {
-      return FindF8SubgraphRecursive(instr->mutable_operand(0), visited_instrs,
-                                     subgraph);
-    } else if (instr->opcode() == HloOpcode::kMultiply) {
-      for (int k = 0; k < 2; ++k) {
-        auto mult_subgraph = FindF8SubgraphRecursive(instr->mutable_operand(k),
-                                                     visited_instrs, subgraph);
-        if (mult_subgraph.has_value()) {
-          return mult_subgraph;
-        }
+  }
+  if (instr->operand_count() == 1 || instr->opcode() == HloOpcode::kDivide ||
+      instr->opcode() == HloOpcode::kDynamicSlice ||
+      instr->opcode() == HloOpcode::kPad) {
+    return FindF8SubgraphRecursive(instr->mutable_operand(0), visited_instrs,
+                                   subgraph);
+  } else if (instr->opcode() == HloOpcode::kMultiply ||
+             instr->opcode() == HloOpcode::kSelect) {
+    for (int k = 0; k < 2; ++k) {
+      int operand_idx = k + (instr->opcode() == HloOpcode::kSelect);
```

Comment:
Add a small comment like "Iterate over operands 0 and 1 for multiply and operands 1 and 2 for select"